code/game/machinery/dehumidifier.dm

```diff
@@ -0,0 +1,161 @@
+		else
+			to_chat(user, "The hatch must be open to insert a power cell.")
+			return
+	else if(isScrewdriver(I))
+		panel_open = !panel_open
+		user.visible_message("<span class='notice'>[user] [panel_open ? "opens" : "closes"] the hatch on the [src].</span>", "<span class='notice'>You [panel_open ? "open" : "close"] the hatch on the [src].</span>")
+		update_icon(1)
+		if(!panel_open && user.machine == src)
+			show_browser(user, null, "window=dehumidifier")
+			user.unset_machine()
+	else
+		..()
+	return
+
+/obj/machinery/dehumidifier/attack_hand(mob/user as mob)
+	..()
+	interact(user)
+
+/obj/machinery/dehumidifier/interact(mob/user as mob)
+
+	if(panel_open)
+
+		var/list/dat = list()
+		dat += "Power cell: "
+		if(cell)
+			dat += "<A href='byond://?src=\ref[src];op=cellremove'>Installed</A><BR>"
+		else
+			dat += "<A href='byond://?src=\ref[src];op=cellinstall'>Removed</A><BR>"
+
+		dat += "Power Level: [cell ? round(cell.percent(),1) : 0]%<BR><BR>"
+
+		var/datum/browser/popup = new(usr, "dehumidifier", "Dehumidifier Control Panel")
+		popup.set_content(jointext(dat, null))
+		popup.set_title_image(usr.browse_rsc_icon(src.icon, "dhum-off"))
+		popup.open()
+	else
+		if(!cell || !cell.charge)
+			on = 0
+			return
+		on = !on
+		user.visible_message("<span class='notice'>[user] switches [on ? "on" : "off"] the [src].</span>","<span class='notice'>You switch [on ? "on" : "off"] the [src].</span>")
+		update_icon()
+	return
+
+
+/obj/machinery/dehumidifier/Topic(href, href_list, state = GLOB.physical_state)
+	if (..())
+		show_browser(usr, null, "window=dehumidifier")
+		usr.unset_machine()
+		return 1
+
+	switch(href_list["op"])
+
+		if("cellremove")
+			if(panel_open && cell && !usr.get_active_hand())
+				usr.visible_message("<span class='notice'>\The usr] removes \the [cell] from \the [src].</span>", "<span class='notice'>You remove \the [cell] from \the [src].</span>")
+				cell.update_icon()
+				usr.put_in_hands(cell)
+				cell.add_fingerprint(usr)
+				cell = null
+				power_change()
+
+		if("cellinstall")
+			if(panel_open && !cell)
+				var/obj/item/weapon/cell/C = usr.get_active_hand()
+				if(istype(C))
+					if(!usr.unEquip(C, src))
+						return
+					cell = C
+					C.add_fingerprint(usr)
+					power_change()
+					usr.visible_message("<span class='notice'>[usr] inserts \the [C] into \the [src].</span>", "<span class='notice'>You insert \the [C] into \the [src].</span>")
+
+	updateDialog()
+
+/obj/machinery/dehumidifier/Process()
+	if(on)
+		if(cell && cell.charge)
+			var/turf/T = get_turf(src)
+			if(!istype(T)) return
+			var/fluid_here = T.get_fluid_depth()
+			if(fluid_here <= 0)
+				cell.use(drainage*1000*CELLRATE)
+				return
+
+			T.remove_fluid(ceil(fluid_here*drainage))
+			cell.use(drainage*5000*CELLRATE)
+			T.show_bubbles()
+			if(world.time > last_gurgle + 80)
+				last_gurgle = world.time
+				playsound(T, pick(SSfluids.gurgles), 50, 1)
+
+		else
+			on = 0
+			power_change()
+		update_icon()
```

Comment:
Why `update_icon()` every Process()? It looks like only power changes actually change the icon.